11x-new-kafka-roller.md

```diff
@@ -0,0 +1,227 @@
+
+Given the number of new features and changes related to KRaft, it is easiest to rewrite it from scratch rather than refactoring the existing component.  Rewriting with a more structured design, the process for evaluating nodes in various states would become more defined and easier to follow.
+
+## Proposal
+
+The objective of this proposal is to introduce a new KafkaRoller with more structured design resembling a finite state machine. The finite station machine design will make it easier to introduce the following features:
+- Ability to roll brokers in parallel to speed up updates in large clusters.
+- After restarting a broker, allow it to lead the partitions it is the preferred leader for. This will help reducing the impact on clients. 
+- Ability to restart controller nodes in parallel when needed, to help recovering controller qourum.
+- Add configurable wait time between restarts to allow slowing down rolling update.
+
+The new roller will have the same behaviour of the current roller but with the additional features above, however, the implementation will be different following the finite state machine design. 
+
+In this proposal term `node` refers to Kafka process that is either controller, broker or combined. When the term `pod` is used, it refers to the Kubernetes `Pod` resource where one of these Kafka processes is running inside.
+
+## State Machine
+
+Kafka nodes will be observed to transition into states and based on these states, actions will be taken on the nodes, which would result in another state transitions. This process will be executed on the nodes in a specific order and will be repeated for each node until the desired state or the maximum number of attempts is reached. The order will depend on which role the node has and what the currently observed state. This will be similar to the order the current roller executes:
+1. unready controllers
+2. ready controllers 
+3. unready combined nodes
+4. ready combined nodes
+5. active controller
+6. unready brokers
+7. ready brokers
+where readiness is determined by Kubernetes Pod status.
+
+If all nodes reached the desired state, the reconciliation will succeed, otherwise it will fail. There are also some conditions that could result in early termination of the process and immediately failing the reconciliation. This will be explained more later. 
+
+Observation of a node is based on different sources such as Kubernetes API, KafkaAgent and Kafka Admin API. These sources will be abstracted so that state machine is not dependent on their specifics as long as it's getting the information it needs. The abstractions also enable much better unit testing.
+
+### States
```

Comment:
These states don't relate to part about preferred partitions, where does that fit in?

Reply:
Ok that makes sense. In that case I think the implementation summary paragraph needs a tweak, I'll suggest an update.